File: Lexer.py
import json
from FA import DFA


EOF = "$"


class Token:
    def __init__(self, tokens):
        self.tokens = tokens
        self.types = [type for type, word in tokens]
        self.words = [word for type, word in tokens]
        self.pos = 0

    def next(self):
        if self.pos >= len(self.tokens):
            return EOF, EOF
        word = self.tokens[self.pos]
        self.pos += 1
        return word

    def peek(self):
        if self.pos >= len(self.tokens):
            return EOF, EOF
        word = self.tokens[self.pos]
        return word

    def get_types(self):
        return self.types[: self.pos] + ["↑"] + self.types[self.pos :]


class CharStream:
    def __init__(self, input_string):
        self.input = input_string + EOF
        self.pos = 0

    def has_next(self):
        return self.pos < len(self.input) - 1

    def next_char(self):
        if self.pos < len(self.input):
            char = self.input[self.pos]
            self.pos += 1
            return char
        else:
            self.pos += 1
            return EOF

    def roll_back(self):
        if self.pos > 0:
            self.pos -= 1


class Lexer:

    def __init__(self, dfa: DFA, config: dict):
        self.S = dfa.S
        self.s0 = dfa.s0
        self.sigma = dfa.sigma
        self.delta = {}
        for s, c, t in dfa.delta:
            self.delta.setdefault(s, {}).setdefault(c, t)
        self.A = dfa.A
        self.Type = config["Type"]
        self.charCat = config["charCat"]
        self.keyword = config["keyword"]
        self.char_stream = None

    def analyze(self, input_string):
        self.char_stream = CharStream(input_string)
        return self

    def next_word(self):
        if not self.char_stream.has_next():
            return EOF, EOF
        state = self.s0
        lexem = ""
        stack = []
        stack.append("bad")
        while state != "err":
            char = self.char_stream.next_char()
            lexem += char
            if state in self.A:
                stack.clear()
            stack.append(state)
            try:
                cat = next(filter(lambda x: char in self.charCat[x], self.charCat.keys()))
            except StopIteration:
                cat = "other"
            if state in self.delta and cat in self.delta[state]:
                state = self.delta[state][cat]
            else:
                state = "err"
        while state not in self.A and state != "bad":
            state = stack.pop()
            lexem = lexem[:-1]
            self.char_stream.roll_back()
        if state in self.A:
            if state in self.Type:
                if lexem not in self.keyword:
                    return self.Type[state], lexem
            return lexem, lexem
        else:
            return EOF, EOF

    def get_token(self):
        tokens = []
        while True:
            token = self.next_word()
            if token == (EOF, EOF):
                break
            if token[0] != "whitespace":
                tokens.append(token)
        return Token(tokens)


if __name__ == "__main__":
    # 使用 DFA 、 charCat分类器 和 状态转移表 构造词法分析器
    dfa = DFA.load("./input/Lexer/Expr_DFA.txt")
    with open("./input/Lexer/Expr_config.json", "r", encoding="utf-8") as f:
        config = json.load(f)
    lexer = Lexer(dfa, config)
    lexer.analyze("a + b * c")
    print(lexer.get_token().tokens)


    # dfa = DFA.load("./input/Lexer/C_DFA.txt")
    # with open("./input/Lexer/C_config.json", "r", encoding="utf-8") as f:
    #     config = json.load(f)
    # lexerExpr = Lexer(dfa, config)
    # with open("./input/Lexer/C_input.txt", "r", encoding="utf-8") as f:
    #     input_string = f.read()
    # lexerExpr.analyze(input_string)
    # print(lexerExpr.get_token().tokens)

File: utils/Lexer.js
import { DFA } from "./FA.js";
import fs from "fs";

const EOF = "$";

class Token {
    constructor(tokens) {
        this.tokens = tokens;
        this.pos = 0;
    }

    /**
     * @returns {[string, string]}
     */
    nextWord() {
        if (this.pos >= this.tokens.length) {
            return [EOF, EOF];
        }
        const word = this.tokens[this.pos];
        this.pos += 1;
        return word;
    }
    /**
     * @returns {[string, string]}
     */
    peekWord() {
        if (this.pos >= this.tokens.length) {
            return [EOF, EOF];
        }
        return this.tokens[this.pos];
    }
    /**
     * @returns {Array<[string, string]>}
     */
    getTokens() {
        return [...this.tokens.slice(0, this.pos).map(([x]) => x), '↑', ...this.tokens.slice(this.pos).map(([x]) => x)];
    }
}

class CharStream {
    constructor(inputString) {
        this.input = inputString + EOF;
        this.pos = 0;
    }

    /**
     * @returns {boolean}
     */
    hasNext() {
        return this.pos < this.input.length - 1;
    }

    /**
     * @returns {string}
     */
    nextChar() {
        if (this.pos < this.input.length) {
            const char = this.input[this.pos];
            this.pos += 1;
            return char;
        } else {
            this.pos += 1;
            return EOF;
        }
    }

    rollBack() {
        if (this.pos > 0) {
            this.pos -= 1;
        }
    }
}

class Lexer {

    /**
     * @param {DFA} DFA实例
     * @param {Object} config 配置文件对象
     * @returns {Lexer} 词法分析器
     */
    constructor(dfa, config) {
        this.S = dfa.S;
        this.s0 = dfa.s0;
        this.delta = {};
        for (let [s, c, t] of dfa.delta) {
            if (!this.delta[s]) this.delta[s] = {};
            this.delta[s][c] = t;
        }
        this.A = dfa.A;
        this.Type = config.Type;
        this.charCat = config.charCat;
        this.keyword = config.keyword;
        this.charStream = null;
    }

    /**
     * 词法分析
     * @param {string} inputString 输入字符串
     * @returns {Lexer} 词法分析器
     */
    analyze(inputString) {
        this.charStream = new CharStream(inputString);
        return this;
    }

    /**
     * @returns {[string, string]} 下一个词
     */
    nextWord() {
        if (!this.charStream.hasNext()) {
            return [EOF, EOF];
        }

        let state = this.s0;
        let lexem = "";
        const stack = ["bad"];

        while (state !== "err") {
            const char = this.charStream.nextChar();
            lexem += char;
            if (this.A.includes(state)) {
                stack.length = 0; // clear stack
            }

            stack.push(state);

            const item = Object.entries(this.charCat).find(([key, value]) =>
                value.includes(char)
            );
            const cat = item ? item[0] : "other";
            if (this.delta[state] && this.delta[state][cat]) {
                state = this.delta[state][cat];
            } else {
                state = "err";
            }
        }

        while (!this.A.includes(state) && state !== "bad") {
            state = stack.pop();
            lexem = lexem.slice(0, -1);
            this.charStream.rollBack();
        }

        if (this.A.includes(state)) {
            if (this.keyword.includes(lexem)) {
                return [lexem, lexem]
            } else if (this.Type[state]) {
                return [this.Type[state], lexem]
            } else {
                return [lexem, lexem]
            }
        } else {
            return [EOF, EOF];
        }
    }

    getToken() {
        const tokens = [];
        while (true) {
            const token = this.nextWord();
            if (token[0] === EOF && token[1] === EOF) {
                break;
            }
            if (token[0] !== "whitespace") {
                tokens.push(token);
            }
        }
        return new Token(tokens);
    }
}
export { Lexer, Token };


// const dfa = DFA.load("./input/Lexer/Expr_dfa.txt");
// const config = JSON.parse(fs.readFileSync("./input/Lexer/Expr_config.json", "utf8"));
// const dfa = DFA.load("./input/Lexer/C_DFA.txt");
// const config = JSON.parse(fs.readFileSync("./input/Lexer/C_config.json", "utf8"));

// const lexer = new Lexer(dfa, config);
// console.table([lexer.S, lexer.s0, lexer.delta, lexer.A]);
// console.log(lexer);
// lexer.analyze('(2 + 3) * a / b + 1');
// console.log(lexer.getToken().tokens);
